### DISCUSSION TITLE
Can shadow simulate large-scale peer-to-peer networks?

### DISCUSSION THREAD

**Discussion by zyansheep** [shadow/shadow#2406] [Q&A] [answered]
I came across shadow by sheer coincidence and after reading a bit of the code and docs I think it is exactly what I've been looking for.

I have a large-scale peer-to-peer [project](https://dither.link/) for which I have been working on a network simulator in rust for almost a year now, but my simulator is real-time and not event-based and thus probably wont be even close to shadow in terms of accuracy.

Would setting up simulations with hundreds of nodes and tens of thousands of connections all with different latencies be even remotely possible using shadow's [rust API](https://shadow.github.io/docs/rust/shadow_rs/core/manager/struct.Manager.html)? I'm hoping to create a GUI application using shadow as the backend and adding features like stopping mid-simulation, saving and loading network states, controling specific nodes mid simulation and other stuff.

**Comment by sporksmith** [comment, marked as answer]
Welcome!

> Would setting up simulations with hundreds of nodes and tens of thousands of connections

In general the only practical limit is the resources on the simulation machine, in particular having enough RAM to run all of your nodes at once. That scale is generally doable, though.

Whether your particular app will work under Shadow largely depends on what system calls it makes and whether Shadow supports them. We'd be interested to hear whether it does and help debug if it doesn't! See also https://shadow.github.io/docs/guide/limitations.html

> all with different latencies be even remotely possible

Configuring a network with different latencies between hosts is supported. See https://shadow.github.io/docs/guide/network_graph_overview.html

> using shadow's [rust API](https://shadow.github.io/docs/rust/shadow_rs/core/manager/struct.Manager.html)?

Shadow is designed to be used an application, not as a library - those documents are intended for developers working on Shadow itself. In particular we don't offer any stability guarantees for using the Rust APIs directly (@robgjansen maybe we should clarify that explicitly on https://shadow.github.io/docs/guide/semver.html). The above should be doable using Shadow as an application, though.

> I'm hoping to create a GUI application using shadow as the backend and adding features like stopping mid-simulation, saving and loading network states, controlling specific nodes mid simulation and other stuff.

Since shadow isn't designed to be used as a library, you'd need to add them to Shadow itself. 

I'll defer to @robgjansen about to what degree we'd be receptive to adding those features or merging PRs for them. In general, Shadow's goal is to be a deterministic simulator, not an interactive debugging/exploration tool. That said, we may be open to adding specific features or hooks to make such things possible. e.g. maybe having a supported way for managed processes to make native syscalls (https://github.com/shadow/shadow/issues/2157) could be used to have your applications communicate directly with a native control process (your GUI), proxy network traffic in/out of the simulation, etc. (Disclaimer: adding such functionality could still be quite tricky)

**Reply by zyansheep** [reply]
Thank You for the warm welcome!

> Shadow is designed to be used an application, not as a library

> what degree we'd be receptive to adding those features or merging PRs for them

Would adding stability guarantees and good documentation to the existing rust API be on the list of PRs you'd be willing to accept? I have this notion that standalone applications are kinda difficult to integrate into existing codebases and it would be 100x easier to just be able to call into a rust API to programmatically manage the execution of the simulator :)

**Reply by robgjansen** [reply]
We're in the middle of a large migration of our code-base from C to Rust, and as you could imagine that is an enormous effort. We are being very cautious at this point with regards to adding large new features that may be difficult to maintain. We will be in a much better position to consider such extensions once we have a stable simulator fully written in Rust.

As @sporksmith mentioned, we are focused on the simulator executable and related code. Creating and maintaining language-specific library bindings seems like a big enough nightmare that we'd want to avoid it, at least for now.

That being said, we have previously considered building a message-passing API that exposes some simulation information and could allow external tools to control the simulator. See #1065. This would be a language-agnostic way to expose simulation internals and would not require maintenance of library bindings in the same was as a library API would.

PRs that improve our documentation are always welcome, even for internal APIs that are only really meant to be read by other people writing Shadow code :)
